feat(cli): support TS/JS config files

dwieeb opened the pull request:
Devs can now have `capacitor.config.ts` (or `capacitor.config.js`) files, for example:

```typescript
import { CapacitorConfig } from '@capacitor/cli';

const config: CapacitorConfig = {
  appId: 'com.capacitorjs.testapp',
  appName: 'capacitor-testapp',
  bundledWebRuntime: false,
  webDir: 'build',
  plugins: {
    SplashScreen: {
      launchShowDuration: 0,
    },
  },
};

export = config;
```

depends on https://github.com/ionic-team/capacitor/pull/3755
resolves https://github.com/ionic-team/capacitor/issues/3141

aparajita commented:
> Still debating if TS config files should be supported

What are the technical issues in implementing that? I assume you would compile on the fly.

dwieeb commented:
> What are the technical issues in implementing that? I assume you would compile on the fly.

We'd have to ship typescript as a dependency and then do something like [what Stencil does](https://github.com/ionic-team/stencil/blob/master/src/compiler/sys/node-require.ts#L15-L61).

aparajita commented:
> We'd have to ship typescript as a dependency and then do something like what Stencil does.

Yeah, I figured it would be something like that. If you already have working code you can steal from Stencil, you might as well do it. You could add TypeScript as an optional dependency.

jcesarmobile commented on cli/src/cordova.ts:
```diff
@@ -527,7 +527,7 @@ export async function getCordovaPreferences(config: Config): Promise<any> {
       `${c.strong(
         `Cordova preferences can be automatically ported to ${c.strong(
-          'capacitor.config.json',
+          config.app.extConfigName,
```
init only works on .json files, don't need to change this

jcesarmobile reviewed (changes requested):
Added a few comments on the definitions.
`ios.scrollEnabled` is missing.

Also, since init won't work if a .ts/.js file is present, don't need to change the code in a few places, I commented the places where it doesn't need to change.

dwieeb commented:
Added `scrollEnabled` (isn't documented currently, btw: https://capacitorjs.com/docs/config)

I reworded the bundle option and clarified the default value for `allowsLinkPreview` per your suggestions 👍 

> Also, since init won't work if a .ts/.js file is present, don't need to change the code in a few places, I commented the places where it doesn't need to change.

I chose not to use the config name in init. I think it might be easier when maintaining it in the future (easier to search for `capacitor.config.json`), but I don't feel strongly about it

> Since there is already a common confusion about the android package name and the application ID, I think we should not use package ID here, better use App ID.

For now I changed it back, but we should probably have a bigger discussion because Ionic, Capacitor, and Appflow all call it different things. 😂 I copied this description from [create-capacitor-app](https://github.com/ionic-team/create-capacitor-app/), but I can change it to App ID there too.

buschtoens commented on cli/src/config.ts:
```diff
@@ -57,6 +54,86 @@ export async function loadConfig(): Promise<Config> {
+      extConfigType: 'ts',
+      extConfigName,
+      extConfigFilePath: extConfigFilePath,
+      extConfig: requireTS(ts, extConfigFilePath) as any,
```
I think it'd be quite useful to optionally support exporting a `() => Promise<ExternalConfig>` instead of just `ExternalConfig`.

Would you accept a pull request?

dwieeb replied:
Yes, I was thinking about this as well. I would be happy to look at a PR that adds that functionality.

webuniverseio commented on cli/src/declarations.ts:
```diff
@@ -1,22 +1,340 @@
+     *
+     * @since 1.0.0
+     */
+    url?: string;
```
Would it be possible to clarify why we say this `This is not intended for use in production`?
You can see under https://github.com/ionic-team/capacitor/discussions/5075 this question was asked. In https://github.com/ionic-team/capacitor/issues/4122 people still use it and it fixes problems like https://github.com/ionic-team/capacitor/issues/1373

webuniverseio replied:
@imhoffd
